test/common/network/dns_impl_test.cc

```diff
@@ -2,9 +2,232 @@
 #include "envoy/network/dns.h"
 
 #include "common/api/api_impl.h"
+#include "common/common/logger.h"
+#include "common/event/dispatcher_impl.h"
+#include "common/network/dns_impl.h"
+
+#include <arpa/nameser.h>
+#include <arpa/nameser_compat.h>
+
+#include "ares.h"
+#include "ares_dns.h"
+#include "event2/buffer.h"
+#include "event2/bufferevent.h"
+#include "event2/event.h"
+#include "event2/listener.h"
 
 namespace Network {
 
+namespace {
+
+// List of IP address (in human readable format).
+typedef std::list<std::string> IpList;
+// Map from hostname to IpList.
+typedef std::unordered_map<std::string, IpList> HostMap;
+
+// Represents a single TestDnsServer query state and lifecycle. This implements
+// just enough of RFC 1035 to handle queries we generate in the tests below.
+class TestDnsServerQuery {
```

Comment:
In general, we are trying to break our dependency on libevent so that eventually it's easier to either use libev, write our own event loop, etc. if we want. We no longer use bufferevent in the prod code. I apologize given how much work you put into this, but please redo this code so that it uses the Envoy interfaces all rooted off of Event::Dispatcher (create listener, create client connections, etc.). You can look at ListenerImplTest to see an example of a a small test that uses real connections. It should be relatively easy to refactor this code around that. Like the FileEvent comment above, it will also get rid of a lot of boiler plate C stuff.